monai/apps/reconstruction/networks/blocks/varnetblock.py

```diff
@@ -0,0 +1,79 @@
+# Copyright (c) MONAI Consortium
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import torch
+import torch.nn as nn
+from torch import Tensor
+
+from monai.apps.reconstruction.networks.nets.utils import sens_expand, sens_reduce
+
+
+class VarNetBlock(nn.Module):
+    """
+    A variational block based on Sriram et. al., "End-to-end variational networks for accelerated MRI reconstruction".
+    It applies data consistency and refinement to the intermediate kspace and combines those results.
+
+    Modified and adopted from: https://github.com/facebookresearch/fastMRI
+
+    Args:
+        refinement_model: the model used for refinement (typically a U-Net but can be any deep learning model
+            that performs well when the input and output are in image domain (e.g., a convolutional network).
+        spatial_dims: is 2 for 2D data and is 3 for 3D data
+    """
+
+    def __init__(self, refinement_model: nn.Module, spatial_dims: int = 2):
+        super().__init__()
+        self.model = refinement_model
+        self.spatial_dims = spatial_dims
+        self.dc_weight = nn.Parameter(torch.ones(1))  # learned scalar as the multiplier of the DC block
+
+        buffer_shape = [1 for _ in range(spatial_dims + 3)]  # 3 denotes the batch, channel, and real/complex dimensions
+        self.register_buffer("zeros", torch.zeros(buffer_shape))
+
+    def soft_dc(self, x: Tensor, ref_kspace: Tensor, mask: Tensor) -> Tensor:
+        """
+        Applies data consistency to input x. Suppose x is some intermediate estimate of the kspace and ref_kspace
```

Comment:
Similar to the previous PR on sensitivity model, here the function `soft_dc` is specific to this `VarNetBlock` so I didn't put it under utils.